library/src/main/scala/scala/scalajs/niocharset/UTF_8.scala

```diff
@@ -228,47 +248,34 @@ private[niocharset] object UTF_8 extends Charset("UTF-8", Array(
     @inline private def isInvalidNextByte(b: Int): Boolean =
       (b & 0xc0) != 0x80
 
+    /** requires passed in bytes to be a valid byte sequence **/
     @inline private def decode2(b1: Int, b2: Int): DecodedMultiByte = {
-      if (isInvalidNextByte(b2))
+      val codePoint = (((b1 & 0x1f) << 6) | (b2 & 0x3f))
+      // By construction, 0 <= codePoint <= 0x7ff < MIN_SURROGATE
+      if (codePoint < 0x80) {
+        // Should have been encoded with only 1 byte
         DecodedMultiByte(CoderResult.malformedForLength(1))
-      else {
-        val codePoint = (((b1 & 0x1f) << 6) | (b2 & 0x3f))
-        // By construction, 0 <= codePoint <= 0x7ff < MIN_SURROGATE
-        if (codePoint < 0x80) {
-          // Should have been encoded with only 1 byte
-          DecodedMultiByte(CoderResult.malformedForLength(1))
-        } else {
-          DecodedMultiByte(codePoint.toChar)
-        }
+      } else {
+        DecodedMultiByte(codePoint.toChar)
       }
     }
 
     @inline private def decode3(b1: Int, b2: Int, b3: Int): DecodedMultiByte = {
-      if (isInvalidNextByte(b2))
+      val codePoint = (((b1 & 0xf) << 12) | ((b2 & 0x3f) << 6) | (b3 & 0x3f))
+      // By construction, 0 <= codePoint <= 0xffff < MIN_SUPPLEMENTARY_CODE_POINT
+      if (codePoint < 0x800) {
+        // Should have been encoded with only 1 or 2 bytes
         DecodedMultiByte(CoderResult.malformedForLength(1))
-      else if (isInvalidNextByte(b3))
-        DecodedMultiByte(CoderResult.malformedForLength(2))
-      else {
-        val codePoint = (((b1 & 0xf) << 12) | ((b2 & 0x3f) << 6) | (b3 & 0x3f))
-        // By construction, 0 <= codePoint <= 0xffff < MIN_SUPPLEMENTARY_CODE_POINT
-        if (codePoint < 0x800) {
-          // Should have been encoded with only 1 or 2 bytes
-          DecodedMultiByte(CoderResult.malformedForLength(1))
-        } else if (codePoint >= MIN_SURROGATE && codePoint <= MAX_SURROGATE) {
-          // It is a surrogate, which is not a valid code point
-          DecodedMultiByte(CoderResult.malformedForLength(3))
-        } else {
-          DecodedMultiByte(codePoint.toChar)
-        }
+      } else if (codePoint >= MIN_SURROGATE && codePoint <= MAX_SURROGATE) {
+        // It is a surrogate, which is not a valid code point
+        DecodedMultiByte(CoderResult.malformedForLength(3))
+      } else {
+        DecodedMultiByte(codePoint.toChar)
       }
     }
 
     @inline private def decode4(b1: Int, b2: Int, b3: Int, b4: Int): DecodedMultiByte = {
-      if (isInvalidNextByte(b2))
-        DecodedMultiByte(CoderResult.malformedForLength(1))
-      else if (isInvalidNextByte(b3))
-        DecodedMultiByte(CoderResult.malformedForLength(2))
-      else if (isInvalidNextByte(b4))
+      if (isInvalidNextByte(b4))
```

Comment:
This last test being alone here is extremely confusing. I suggest moving it to the two call sites, so that it is uniform wrt. the other bytes.